### PR TITLE
Expose a Decimal32Dtype in cuDF Python

### PR DESCRIPTION
**Pull request by skirui-source**
Fixes: #8218

Similarly to libcudf's 64-bit decimal type,  this PR exposes the `Decimal32Dtype `and its corresponding `Decimal32Column` type.  Following this implementation, user can create a series or dataframe with `decimal32` dtype. 
Note: Only `to_arrow` and `from_arrow` methods are currently being supported. 

**Example:** 
```
>>> import cudf
>>> s = cudf.Series([1,2,3,4], dtype=cudf.Decimal32Dtype(precision=8, scale=2))
>>> s
0    1.00
1    2.00
2    3.00
3    4.00
dtype: decimal32
```

### CODE REVIEW

File: python/cudf/cudf/tests/test_decimal.py [hunk 1]
```diff
@@ -18,39 +18,51 @@
...
+
+def test_from_arrow_max_precision_decimal64():
+    with pytest.raises(ValueError):
+        Decimal64Column.from_arrow(
+            pa.array([1, 2, 3], type=pa.decimal128(scale=0, precision=19))
+        )
 
 
-def test_from_arrow_max_precision():
+def test_from_arrow_max_precision_decimal32():
```
Comment (isVoid) [inline python/cudf/cudf/tests/test_decimal.py:62]: We probably shouldn't rename this test to `decimal32` but `decimal64` and create the 32bit counter parts?

**Review by isVoid — COMMENTED** [review]
Mostly looks good aside one small question. Great job!

File: python/cudf/cudf/tests/test_decimal.py (hunk 1 above)
Comment (skirui-source) [reply python/cudf/cudf/tests/test_decimal.py:62]: Actually I have both tests as follows, please confirm?

Line 55-59: `def test_from_arrow_max_precision_decimal64():`
Line 62-66: `def test_from_arrow_max_precision_decimal32():`

Comment (isVoid) [reply python/cudf/cudf/tests/test_decimal.py:62]: Oops. yes they exists.

**Review by isVoid — APPROVED** [review]

File: python/cudf/cudf/core/column/column.py [hunk 2]
```diff
@@ -973,7 +975,7 @@ def as_string_column(
 
     def as_decimal_column(
```
Comment (charlesbluca) [inline python/cudf/cudf/core/column/column.py:976]: Would we want to split `as_decimal_column` off into `as_decimal32_column` and `as_decimal64_column` where it's implemented?

**Review by charlesbluca — COMMENTED** [review]
Looks great 🙂  just one small question

File: python/cudf/cudf/utils/dtypes.py [hunk 3]
```diff
@@ -158,6 +159,170 @@ def numeric_normalize_types(*args):
     return [a.astype(dtype) for a in args]
 
 
+def is_numerical_dtype(obj):
```
Comment (shwina) [inline python/cudf/cudf/utils/dtypes.py:162]: I believe all these functions got moved to `cudf/api/types/`. Looks like some git mishap led to these changes.

Comment (shwina) [reply python/cudf/cudf/utils/dtypes.py:162]: To fix, you should be able to:

1. Move the functions `is_decimal_*` to `cudf/api/types.py`.
2. Undo all the changes to this file (i.e., delete all this functions, and undo the import changes on line 11 and 12).

File: python/cudf/cudf/core/dtypes.py [hunk 4]
```diff
@@ -243,9 +243,110 @@ def __hash__(self):
...
+    def _from_decimal(cls, decimal):
+        """
+        Create a cudf.Decimal32Dtype from a decimal.Decimal object
+        """
+        metadata = decimal.as_tuple()
+        precision = max(len(metadata.digits), -metadata.exponent)
+        return cls(precision, -metadata.exponent)
+
+    def serialize(self) -> Tuple[dict, list]:
+        return {"precision": self.precision, "scale": self.scale}, []
```
Comment (charlesbluca) [inline python/cudf/cudf/core/dtypes.py:340]: Depending on if #8441 gets merged first, we would also want to add `type-serialized` to the header here, which would be used in serialization of list/struct dtypes

Comment (charlesbluca) [inline python/cudf/cudf/core/dtypes.py:340]:
```suggestion
        return (
            {
                "type-serialized": pickle.dumps(type(self)),
                "precision": self.precision,
                "scale": self.scale,
            },
            [],
        )
```

To address #8441 getting merged

File: python/cudf/cudf/core/column/decimal.py [hunk 5]
```diff
@@ -98,7 +152,11 @@ def binary_operator(self, op, other, reflect=False):
         elif op in ("eq", "ne", "lt", "gt", "le", "ge"):
             if not isinstance(
                 other,
-                (DecimalColumn, cudf.core.column.NumericalColumn, cudf.Scalar),
+                (
+                    Decimal64Dtype,
```
Comment (shwina) [inline python/cudf/cudf/core/column/decimal.py:156]:
```suggestion
                    Decimal64Column,
```

I think this is why some of the binaryop tests are failing.

**Review by isVoid — APPROVED** [review]
lgtm code wise. Great job!

Can you also update this page to include the new datatype? https://docs.rapids.ai/api/cudf/stable/basics.html

**Comment by skirui-source** [comment]
@gpucibot merge

**Pull request merged**